File: src/hey/LongestSubstringWithoutRepeatingCharacters.py
# problem: Longest Substring Without Repeating Characters
# author: everpuck
# date: 2/2/2019


class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        sub_list = []
        ret = 0
        for c in s:
            if c in sub_list:
                ret = max(len(sub_list), ret)
                sub_list = sub_list[sub_list.index(c) + 1: ]
                sub_list.append(c)
            else:
                sub_list.append(c)
        
        ret = max(len(sub_list), ret)
        return ret


if __name__ == '__main__':
    s = 'abcabcbb'
    s = 'pwwkew'
    s = 'bbb'
    s = 'dvdf'
    solution = Solution()
    print(solution.lengthOfLongestSubstring(s))

File: src/hey/LongestSubstringWithoutRepeatingCharacters.java
/*
    problem: Median of Two Sorted Arrays
    author: everpuck
    date: 02/09/2019
*/


class Solution {
    public int lengthOfLongestSubstring(String s) {
        String tempString = "";
        int ret = 0;
        char c;
        for (int i = 0; i < s.length(); ++i) {
            c = s.charAt(i);
            if (tempString.length() > 0 && 
                tempString.indexOf(c) > -1) {
                ret = Math.max(tempString.length(), ret);
                tempString = tempString.substring(tempString.indexOf(c) + 1);
                tempString += c;
            }
            else {
                tempString += c;
            }
        }
        ret = Math.max(tempString.length(), ret);
        return ret;
    }

    /* Official sliding window algrorithm
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        Set<Character> set = new HashSet<>();
        int ans = 0, i = 0, j = 0;
        while (i < n && j < n) {
            // try to extend the range [i, j]
            if (!set.contains(s.charAt(j))){
                set.add(s.charAt(j++));
                ans = Math.max(ans, j - i);
            }
            else {
                set.remove(s.charAt(i++));
            }
        }
        return ans;
    }
    */

}


public class LongestSubstringWithoutRepeatingCharacters {
    public static void main(String[] args) {
        int ret = 0;
        String s = new String("cdd");
        Solution solution = new Solution();
        ret = solution.lengthOfLongestSubstring(s);
        System.out.println(" ret: " + ret);
    }
}